[Migrate] From `EventBus` to `Coroutines`

ParaskP7 opened the pull request:
Closes: [AINFRA-749](https://linear.app/a8c/issue/AINFRA-749)

---

### Description

This PR is a prerequisite for [AINFRA-750](https://linear.app/a8c/issue/AINFRA-750/replace-wellsql-with-room) (`WellSQL` -> `Room`), which prepares the ground for this final migration. Main changes include but not limited to:

- Removing all `wordpress-fluxc` related dependencies, then replacing `EventBus` with `Coroutines`. Also, removing any `FluxC` related references from the library.
- Making the library easier to use for clients.
- Making `EncryptedLogStore` a singleton.
- Fixing `EncryptedLogTest` by replacing `CountDownLatch` with `Turbine`.

---

### Testing Steps

Using the [DOAndroid](https://github.com/bloom/DayOne-Android) or/and [PCAndroid](https://github.com/Automattic/pocket-casts-android) project, update `automattic-encryptedlogging` to `19-ed6bf1f0a0637dec7da8c366813400485a96f1f2` (version which is based on this PR's [latest](https://github.com/Automattic/EncryptedLogging/pull/19/commits/ed6bf1f0a0637dec7da8c366813400485a96f1f2) commit) and make sure that encrypted logging continues to work as expected:

1. Using `App Inspection` from AS find the `encrypted-logging.db` and its sole `EncryptedLogModel` table. Notice that the table is empty, there are no log to be sent.
2. Make the app crash. Using [DOAndroid](https://github.com/bloom/DayOne-Android) for example, you could navigate to `Settings` -> `Developer` and then click `Crash The App`.
3. You might now notice a new db entry within `App Inspection` from AS on that `DETACHED` app (or not, it depends). Now, switch to the `ATTACHED` app and notice the db empty. This is because this db entry/log has been already sent to Sentry already, on app start and after the app crashed. To really notice the entry, trying crashing the app again, this time put the device on offline mode first. Doing that you'll be able to see the entry on the db, even when switching to the newly created `ATTACHED` app.
4. Open [Sentry](https://a8c.sentry.io) and navigate to `Issues`. Find the project you're testing for (for example `day-one-android` or/and `pocket-casts-android`) and check the latest `debug` exceptions. Using [DOAndroid](https://github.com/bloom/DayOne-Android) for example, you could search for `Developer caused a crash!` and check the latest crash. Now navigate at the bottom and check the `uuid`. Verify it is the same `UUID` that you saw on that db entry before.
5. Open [Read Encrypted Logs](https://mc.a8c.com/encrypted-logs.php), copy-paste that `UUID` into `Log UUID`, click `View` and make sure you can read the newly uploaded and now decrypted log.

---

Additionally, run this [ReleaseStack_EncryptedLogTest](https://github.com/Automattic/EncryptedLogging/blob/trunk/encryptedlogging/src/androidTest/kotlin/com/automattic/encryptedlogging/release/ReleaseStack_EncryptedLogTest.kt#L48)  instrumented test locally, and make sure that it runs successfully, and as expected. Note that:

1. It is quite hard to work with this test because `TooManyRequests` can be hit quite easily. At which point you need to wait for a few minutes (15' minutes to half an hour) before being able to verify it is working as expected.
2. In order to verify that this test works as expected one would need to use the debugger and navigate through the flow manually. Otherwise the test might success with a false positive, but actually silently failing on (at least) this `TooManyRequests` problem mentioned above.

ParaskP7 committed:
Migrate: Remove wordpress-fluxc dependencies

ParaskP7 committed:
Test: Add kotlinx-coroutines-test dependency for android test to compile

ParaskP7 committed:
Migrate: Replace eventbus with coroutines

ParaskP7 committed:
Refactor: Remove suspect to make sdk easier to use for clients

PR Discussion: https://github.com/Automattic/EncryptedLogging/pull/17
#discussion_r2120767596

ParaskP7 committed:
Test: Reformat encrypted log test by splitting long lines

ParaskP7 committed:
Test: Replace deprecated junit assertions with assertj

------------------------------------------------------------------------

Also, as part of this change the 'packaging' configuration got added due
to the fact that 'assertj-core' and 'jspecify' include conflicting files
at 'META-INF/versions/9/OSGI-INF/MANIFEST.MF', see build failure below:

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task
 ':encryptedlogging:mergeDebugAndroidTestJavaResource'.
> A failure occurred while executing com.android.build.gradle.internal.
 tasks.MergeJavaResWorkAction
   > 2 files found with path 'META-INF/versions/9/OSGI-INF/MANIFEST.MF'
    from inputs:
      - org.assertj:assertj-core:3.27.3/assertj-core-3.27.3.jar
      - org.jspecify:jspecify:1.0.0/jspecify-1.0.0.jar
     Adding a packaging block may help, please refer to
     https://developer.android.com/reference/tools/gradle-api/com/
      android/build/api/dsl/Packaging
     for more information

ParaskP7 committed:
Test: Replace deprecated kotlin create temp file with java

Explanation: Avoid creating temporary files in the default temp location
with this function due to too wide permissions on the newly created
file. Use 'kotlin.io.path.createTempFile' instead or resort to
'java.io.File.createTempFile'.

ParaskP7 committed:
Test: Fix encrypted log test by replacing count down latch with turbine

FYI.1: It is quite hard to work with this test because 'TooManyRequests'
can be hit quite easily, at which point you need to wait for a few
minutes (15' minutes to half an hour) before being able to verify that
the test is working as expected.

FYI.2: Also, in order to verify that the test is working as expected one
would need to use the debugger and navigate through the flow manually.
Otherwise the test might success with a false positive, but actually
silently failing on (at least) this 'TooManyRequests' problem mentioned
above.

ParaskP7 committed:
Refactor: Rename package-name fluxc to encrypted-log preferences

This is done so that 'fluxc' as a keyword is removed from everywhere
within this library (but the README, which mentions its origin).

PS: Also, as part of this change the 'getFluxCPreferences(...)'
functions got rename to 'getPreferences', for the very same reason.

ParaskP7 committed:
Refactor: Automatically reformat encrypted log store

ParaskP7 committed:
Refactor: Automatically reformat encrypted log test

ParaskP7 committed:
Refactor: Make encrypted log store a singleton

PS Discussion: https://github.com/Automattic/EncryptedLogging/pull/17
#discussion_r2120760639

ParaskP7 committed:
Analysis: Suppress unused for parameter e of unsatisfied link error

ParaskP7 committed:
Analysis: Use the ktx extension function shared preferences edit

Inspection info: The Android KTX libraries decorates the Android
platform SDK as well as various libraries with more convenient extension
functions available from Kotlin, allowing you to use default parameters,
named parameters, and more.

ParaskP7 committed:
Analysis: Resolve max cache size in bytes magic number

ParaskP7 commented on encryptedlogging/src/main/kotlin/com/automattic/encryptedlogging/store/EncryptedLogStore.kt:
```diff
@@ -52,67 +43,66 @@ private const val MAX_RETRY_COUNT = 3
 
-    init {
-        WellSql.init(encryptedWellConfig)
+        fun getInstance(
```
FYI: I actually considered that but ended-up deciding to keep it simple, with a `getInstance(...)` method that is not synchronized.

ParaskP7 commented on encryptedlogging/src/main/kotlin/com/automattic/encryptedlogging/AutomatticEncryptedLogging.kt:
```diff
@@ -73,18 +68,24 @@ public class AutomatticEncryptedLogging(
+        CoroutineScope(Dispatchers.IO).launch {
+            encryptedLogStore.queueLogForUpload(payload)
+        }
     }
 
-    override suspend fun uploadEncryptedLogs() {
-        encryptedLogStore.uploadQueuedEncryptedLogs()
+    override fun uploadEncryptedLogs() {
+        CoroutineScope(Dispatchers.IO).launch {
+            encryptedLogStore.uploadQueuedEncryptedLogs()
+        }
     }
 
     override fun resetUploadStates() {
-        dispatcher.dispatch(EncryptedLogActionBuilder.newResetUploadStatesAction())
+        CoroutineScope(Dispatchers.IO).launch {
```
Copilot, can you suggest the code change you have in mind? 🙏

wzieba replied:
It's a pity, but it seems it does't support replies :( 

> Copilot's review comments behave like review comments from humans. You can add reactions to them, comment on them, resolve them and hide them.
Any comments you add to Copilot's review comments will be visible to humans, but they won't be visible to Copilot, and Copilot won't reply.

ParaskP7 replied:
Yea, wanted to try anyway... wishful thinking! 🫣 😅 😞

wzieba replied:
The comment is valid though: WDYT about creating a single scope? Something like this: 
https://github.com/Parsely/parsely-android/blob/d2c189699150943f884cfc62d904026b7ed3f210/parsely/src/main/java/com/parsely/parselyandroid/ParselyCoroutineScope.kt#L8-L9

ParaskP7 replied:
Awesome, love it, this is now done @wzieba and thanks for the suggestion you both (😄): 1080d856d4e7d37c275db2755df5541aae70e474

wzieba commented on encryptedlogging/src/androidTest/kotlin/com/automattic/encryptedlogging/release/ReleaseStack_EncryptedLogTest.kt:
```diff
@@ -61,93 +63,121 @@ internal class ReleaseStack_EncryptedLogTest {
-        initializeEncryptedLogStore(context, preferenceUtilsWrapper)
-        WellSql.delete(EncryptedLogModel::class.java).execute()
-        mDispatcher.register(this)
+        Dispatchers.setMain(testDispatcher)
```
What's the reason for modifying the main dispatcher? I think this SDK shouldn't do anything on the main thread, so this looks redundant. I propose removing it, here's my idea: c686859302f7aab7c8b05602ea576c33fd058c6a. Tests passed.

ParaskP7 replied:
Oh true, you're totally right @wzieba , I guess this a habit of working on an Android codebase and writing tests for `ViewModel` classes, plus AI suggesting it, done: 3eac8a20f3d32ee16f3e512dcb9a3840636a57c5

wzieba reviewed (commented):
Looks good! Added some suggestions

ParaskP7 committed:
Test: Remove unnecessary test dispatcher main configuration from test

PR Comment: https://github.com/Automattic/EncryptedLogging/pull/19
#discussion_r2154699460

ParaskP7 committed:
Refactor: Define and reuse dedicated scope for launching sdk coroutines

PR Comment: https://github.com/Automattic/EncryptedLogging/pull/19
#discussion_r2154708447

wzieba reviewed (approved):
Thanks for applying changes! :shipit:
